pydantic_extra_types/timezone_name.py

```diff
@@ -0,0 +1,125 @@
+"""Time zone name validation and serialization module."""
+
+from __future__ import annotations
+
+import importlib
+import sys
+import warnings
+from typing import Any, List
+
+from pydantic import GetCoreSchemaHandler, GetJsonSchemaHandler
+from pydantic_core import PydanticCustomError, core_schema
+
+
+def _is_available(name: str) -> bool:
+    try:
+        importlib.import_module(name=name)
+        return True
+    except ModuleNotFoundError:  # pragma: no cover
+        return False
+
+
+if _is_available('zoneinfo') and _is_available('tzdata'):  # pragma: no cover
+    from zoneinfo import available_timezones
+
+    def _tz_provider() -> set[str]:
+        return set(available_timezones())
+
+elif _is_available('pytz'):  # pragma: no cover
+    if sys.version_info[:2] > (3, 8):
+        warnings.warn(
+            'Projects using Python 3.9 or later'
+            ' should be using the support now included as part of the standard library zone-info. '
+            'Please consider switching to the standard library module.'
+        )
+    from pytz import all_timezones
+
+    def _tz_provider() -> set[str]:
+        return set(all_timezones)
+else:  # pragma: no cover
+    if sys.version_info[:2] == (3, 8):
+        raise ImportError('No pytz module not found. Please install it with "pip install pytz')
+    raise ImportError('No timezone provider found. Please install tzdata' 'Please install it with "pip install tzdata"')
+
```

Comment:
I was thinking on something like this:

```py
def _is_available(name: str) -> bool:
    """Check if a module is available for import."""
    try:
        importlib.import_module(name)
        return True
    except ModuleNotFoundError:
        return False

def _tz_provider_from_zoneinfo() -> set[str]:
    """Get timezones from the zoneinfo module."""
    from zoneinfo import available_timezones
    return set(available_timezones())

def _tz_provider_from_pytz() -> set[str]:
    """Get timezones from the pytz module."""
    from pytz import all_timezones
    return set(all_timezones)

def _warn_about_pytz_usage():
    """Warn about using pytz with Python 3.9 or later."""
    warnings.warn(
        'Projects using Python 3.9 or later should be using the support now included as part of the standard library zone-info. '
        'Please consider switching to the standard library module.'
    )

def get_timezone_provider() -> set[str]:
    """Determine the timezone provider and return available timezones."""
    if _is_available('zoneinfo') and _is_available('tzdata'):
        return _tz_provider_from_zoneinfo()
    elif _is_available('pytz'):
        if sys.version_info[:2] > (3, 8):
            _warn_about_pytz_usage()
        return _tz_provider_from_pytz()
    else:
        if sys.version_info[:2] == (3, 8):
            raise ImportError('No pytz module found. Please install it with "pip install pytz"')
        raise ImportError('No timezone provider found. Please install tzdata with "pip install tzdata"')
```

and we can next call in the `set()` the `get_timezone_provider`

Reply:
simply genius but i consider get_timezones to be sligly better name instead of get_timezone_provider so i used that name instead